### ISSUE TITLE
Empty OneToMany when overflowing LazyLoadBatchSize and mixed queries

### ISSUE THREAD

**Issue by AntoineDuComptoirDesPharmacies** [ebean-orm/ebean#3381]
##  Context
We never had this problem in 4 years. It start doing this since we upgrade from Ebean 13.22.0 to 13.26.1.
We did this upgrade in order to fix the following issue : https://github.com/ebean-orm/ebean/issues/3319

We tried to upgrade to 15.0.2 but the issue still exists.

## Expected behavior

The linked entites should be loaded from the database.

## Actual behavior

We randomly get an empty array when our process access a oneToMany lazy-loaded attribute.
However, the associated entities are well stored in the database and should be returned.

Note : We did not succeed ot reproduce on local development, but this happens once every two days on our production server.

### Steps to reproduce

Order entity :
```java
@Entity
public class Order [...] {
    [...]
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "order", fetch = FetchType.LAZY)
    @Column(nullable = true)
    private List<OrderLine> orderLines;
    [...]
} 
```
OrderLine entity :
```java
public class OrderLine [...] {
  [...]
    @ManyToOne
    @Column(nullable = true)
    private Order order;
  [...]
}
```
Business code that produce the problem :
```java
        try (Transaction transaction = DB.beginTransaction()) {
            elementsToProcess = Order.find
                    .query()
                    .forUpdate()
                    .where()
                    .isNotNull("dirty")
                    .orderById(true)
                    .setMaxRows(15)
                    .findList();

            process(elementsToProcess.stream().toArray(Order[]::new));

            transaction.commit();
        }
```
```java
    public void process(Order... orders){
        try (Transaction txn = DB.beginTransaction()) {
            for (Order order: orders) {
                UpdateQuery query = DB.update(Order.class);
                
                [...]
                

	        BigDecimal totalPriceHt = order.getOrderLines() // HERE, THE LINES ARE EMPTY
                                                          .stream()
                                                          .map(PaymentHelper::getTotalPriceHT)
                                                          .reduce(BigDecimal.ZERO, BigDecimal::add);

                query.set("dirty", null)
                        .set("totalPriceHt", totalPriceHt);

                query.where()
                    .eq("id", order.getId())
                    .eq("dirty", order.getDirty())
                    .update(txn);
            }

            txn.commit();
        }
    }
```

```console
No specific console, everything is running fine, no error, no warn.
```

Thanks in advance for your help.
Yours faithfully,
LCDP

**Comment by AntoineDuComptoirDesPharmacies** [comment]
We discovered something !
We successfuly reproduced the problem. It occurs for all `Order` that are processed above the `10th` Order of the array.

We noticed that Lazy Loading in Ebean have a default value of '10'. May it be correlated ?
[Ebean Lazy Loading](https://ebean.io/docs/query/background/nplus1)

We are going to try to change the batch size value and see if its change the cursor in our bug.

Yours faithfully,
LCDP

**Comment by rob-bygrave** [comment]
> seems to be linked to MixedQuery

What is MixedQuery?

**Comment by rob-bygrave** [comment]
You are producing a failing test case right?

**Comment by AntoineDuComptoirDesPharmacies** [comment]
> > seems to be linked to MixedQuery
> 
> What is MixedQuery?

Maybe we are wrong about the term used, it was a reference to the `setFlushOnMixedQuery` method  but it may not be linked to the flush part because setting `True` or `False` on this method do not change anything.

The problem seems to occur when we execute the update at the end of each iteration, and so we have a mixing of `Update` query and `Select` (lazy loading) query.
When we store all the `UpdateQuery` in an array and execute them all together at the end, we do not have the problem of empty array as a result of lazy loading.

> Why is the nested beginTransaction() in here? We absolutely need the transaction to wrap the database locks (forUpdate) so the nested beginTransaction() should only be here is this method has multiple entry points and other entry points don't have a transaction?

This lock is only a helper on our side to avoir having multiple process executing the method concurrently but the problem still occur even if we remove it.
The nested `beginTransaction()` is certainly not useful here. Only to be sure that a transaction is started during the whole execution of this method. We could have use `TxScope.mandatory()`. We tried to remove this nested transaction and the behavior is unchanged.

> You are producing a failing test case right?
Yes, we can reproduce 100% time, do you have a mini project (like a Codesanbox) where I can try to insert the failing code ?

**Comment by rob-bygrave** [comment]
Example minimal - https://github.com/ebean-orm-examples/example-minimal

**Comment by rob-bygrave** [comment]
> query. forUpdate()

Just a note in case I forget to mention it later, that the `forUpdate()` very likely should be `forUpdateSkipLocked()` - at least for Postgres, Oracle, Hana, and SQL Server.

**Comment by AntoineDuComptoirDesPharmacies** [comment]
> > query. forUpdate()
> 
> Just a note in case I forget to mention it later, that the `forUpdate()` very likely should be `forUpdateSkipLocked()` - at least for Postgres, Oracle, Hana, and SQL Server.

What is the concern with `forUpdate()` compared to `forUpdateSkipLocked` ? Is there something special aside from the potential infinite lock ?

Here is the reproducer : 
[Branch where the Asserts are correct, this branch use Ebean 13.22.0](https://github.com/AntoineDuComptoirDesPharmacies/issue-3381/tree/reproducer/issue-3381-working)

[Branch where the Asserts fail on fourth element, considering lazyLoading is set to '3'](https://github.com/AntoineDuComptoirDesPharmacies/issue-3381/tree/reproducer/issue-3381-working)

**Comment by rbygrave** [comment]
> upgrade from Ebean 13.22.0 to 13.26.1

Both of these versions do not have this issue/bug.


> We did this upgrade in order to fix the following issue : https://github.com/ebean-orm/ebean/issues/3320

This doesn't make sense to me in that for #3320 there was _NO CHANGE_ made to ebean at all. That was more that the application needed to change wrt use of transactions in BeanAdapter. So nothing in 3320 requires an upgrade per se.


> We tried to upgrade to 15.0.2 but the issue still exists.

Now this issue we are hitting here was actually **_introduced_** in versions **_14.0.0 / 15.0.0_** by https://github.com/ebean-orm/ebean/issues/3295 in version . This issue didn't exist prior to that.

**Comment by rbygrave** [comment]
FYI: The fix for this is available in versions 14.0.2.1, 14.0.2.1-javax as patch to 14.0.2 and in versions 14.1.0, 14.1.0-javax, 15.1.0 (which include other changes that has already been merged into master).
